File: backend/app/schemas/scanner.py
from enum import Enum
from typing import List
from typing import Optional
from typing import Union

from pydantic import BaseModel


class Operator(str, Enum):
    eq = "="
    gt = ">"
    lt = "<"
    gte = ">="
    lte = "<="
    ne = "!="


class LogicalOperator(str, Enum):
    AND = "AND"
    OR = "OR"


class SortDirection(str, Enum):
    asc = "asc"
    desc = "desc"


class ScannerFilter(BaseModel):

    field: str

    operator: Operator

    value: Union[float, int, str]


class SortRequest(BaseModel):

    field: str

    direction: SortDirection = SortDirection.desc


class ScannerRequest(BaseModel):

    condition: LogicalOperator = LogicalOperator.AND

    filters: List[ScannerFilter]

    sort: Optional[SortRequest] = None

    page: int = 1

    page_size: int = 50

File: frontend/src/types/scanner.ts
export type LogicalOperator = "AND" | "OR";

export type ComparisonOperator =
  | "="
  | ">"
  | "<"
  | ">="
  | "<="
  | "!=";

export interface ScannerFilter {
  field: string;
  operator: ComparisonOperator;
  value: string | number;
}

export interface SortRequest {
  field: string;
  direction: "asc" | "desc";
}

export interface ScannerRequest {
  condition: LogicalOperator;
  filters: ScannerFilter[];
  sort?: SortRequest;
  page: number;
  page_size: number;
}

export interface ScanResult {
  symbol: string;
  close: number;
  volume: number;
  rsi: number;
  ema20: number;
  ema50: number;
  sma20: number;
  macd: number;
}

export interface ScannerResponse {
  total: number;
  page: number;
  page_size: number;
  results: ScanResult[];
}
